src/iface-opc-ua.c

```diff
@@ -134,94 +136,145 @@ static UA_StatusCode setup_read_data(UA_Variant *value,
 			 data_item->identifier);
 	rc = UA_Client_readValueAttribute(opc_ua_client,
 						  *parent, value);
-
 	free(parent);
 	return rc;
 }
 
-static UA_StatusCode get_type_int(union opc_ua_types *tmp,
-				  void * data,
+static void get_type_int(union opc_ua_types *tmp,
+				  UA_Variant *data,
 				  int value_type_size)
 {
-	UA_StatusCode rc = UA_STATUSCODE_GOOD;
-
-	if (value_type_size == 8)
-		tmp->val_i8 = *(UA_Byte *)data;
-	else if (value_type_size == 16)
-		tmp->val_i16 = *(UA_Int16 *)data;
-	else if (value_type_size == 32)
-		tmp->val_i32 = *(UA_Int32 *)data;
-	else
-		rc = UA_STATUSCODE_BAD;
-
-	return rc;
+	if (value_type_size == 8) {
+		if (UA_Variant_hasScalarType(data,
+				&UA_TYPES[UA_TYPES_SBYTE])) {
+			tmp->val_i8 = *(UA_SByte *)data->data;
+		}
+	} else if (value_type_size == 16) {
+		if (UA_Variant_hasScalarType(data,
+				&UA_TYPES[UA_TYPES_INT16])) {
+			tmp->val_i16 = *(UA_Int16 *)data->data;
+		}
+	} else if (value_type_size == 32) {
+		if (UA_Variant_hasScalarType(data,
+				&UA_TYPES[UA_TYPES_INT32])) {
+			tmp->val_i32 = *(UA_Int32 *)data->data;
+		}
+	} else if (value_type_size == 64) {
+		if (UA_Variant_hasScalarType(data,
+				&UA_TYPES[UA_TYPES_INT64])) {
+			tmp->val_u64 = *(UA_Int64 *)data->data;
+		}
+	}
 }
 
-static UA_StatusCode get_type_uint(union opc_ua_types *tmp,
-				   void * data,
+static void get_type_uint(union opc_ua_types *tmp,
+				   UA_Variant *data,
 				   int value_type_size)
 {
-	UA_StatusCode rc = UA_STATUSCODE_GOOD;
+	if (value_type_size == 8) {
```

Comment:
I would recommend replacing the number 8 with something like a name with more details: size_val_8.